src/aliased_buffer.h

```diff
@@ -0,0 +1,191 @@
+      v8::Isolate* isolate,
+      const size_t byte_offset,
+      const size_t count,
+      const AliasedBuffer<uint8_t,
+      v8::Uint8Array>& backing_buffer) :
+      isolate_(isolate),
+      count_(count),
+      byte_offset_(byte_offset),
+      free_buffer_(false) {
+    const v8::HandleScope handle_scope(isolate_);
+    // validate that the byte_offset is aligned with sizeof(NativeT)
+    CHECK_EQ(byte_offset & (sizeof(NativeT) - 1), 0);
+    buffer_ = reinterpret_cast<NativeT*>(
+        const_cast<uint8_t*>(backing_buffer.GetNativeBuffer() + byte_offset));
+
+    v8::Local<v8::ArrayBuffer> ab = backing_buffer.GetArrayBuffer();
+    v8::Local<V8T> js_array = V8T::New(ab, byte_offset, count);
+    js_array_ = v8::Global<V8T>(isolate, js_array);
+  }
+
+  AliasedBuffer(const AliasedBuffer& that) :
+      isolate_(that.isolate_),
+      count_(that.count_),
+      byte_offset_(that.byte_offset_),
+      buffer_(that.buffer_),
+      free_buffer_(false) {
+    js_array_ = v8::Global<V8T>(that.isolate_, that.GetJSArray());
+  }
+
+  ~AliasedBuffer() {
+    if (free_buffer_ && buffer_ != NULL) {
+      free(buffer_);
+    }
+    js_array_.Reset();
+  }
+
+  /**
+   * Helper class that is returned from operator[] to support assignment into
+   * a specified location.
+   */
+  class Reference {
+   public:
+    Reference(AliasedBuffer<NativeT, V8T>* aliased_buffer, size_t index) :
+      aliased_buffer_(aliased_buffer),
+      index_(index) {
+    }
+
+    Reference(const Reference& that) :
+        aliased_buffer_(that.aliased_buffer_),
+      index_(that.index_) {
+    }
+
+    inline Reference& operator=(const NativeT &val) {
+      aliased_buffer_->SetValue(index_, val);
+      return *this;
+    }
+
+    operator NativeT() const {
+      return aliased_buffer_->GetValue(index_);
+    }
+
+   private:
+    AliasedBuffer<NativeT, V8T>* aliased_buffer_;
+    size_t index_;
+  };
+
+  /**
+   *  Get the underlying v8 TypedArray overlayed on top of the native buffer
+   */
+  v8::Local<V8T> GetJSArray() const {
+    return js_array_.Get(isolate_);
+  }
+
+  /**
+  *  Get the underlying v8::ArrayBuffer underlying the TypedArray and
+  *  overlaying the native buffer
+  */
+  v8::Local<v8::ArrayBuffer> GetArrayBuffer() const {
+    return GetJSArray()->Buffer();
+  }
+
+  /**
+   *  Get the underlying native buffer. Note that all reads/writes should occur
+   *  through the GetValue/SetValue/operator[] methods
+   */
+  inline const NativeT* GetNativeBuffer() const {
+    return buffer_;
+  }
+
+  /**
+   *  Synonym for GetBuffer()
+   */
+  inline const NativeT* operator * () const {
+    return GetNativeBuffer();
+  }
+
+  /**
+   *  Set position index to given value.
+   */
+  inline void SetValue(const size_t index, NativeT value) {
+    CHECK_LT(index, count_);
+    buffer_[index] = value;
+  }
+
+  /**
+   *  Get value at position index
+   */
+  inline const NativeT GetValue(const size_t index) const {
+    CHECK_LT(index, count_);
```

Comment:
@trevnorris I’m pretty sure the reported performance difference is due to this check, it’s the only visible difference in the generated output. I’d suggest wrapping it in `#if defined(DEBUG) && DEBUG`?

Reply:
confirmed. that is where the performance hit is coming from. wrapping it sounds good

Reply:
this and the SetValue check are now in debug-only builds.